docs/notebooks/deep_gaussian_processes.pct.py

```diff
@@ -0,0 +1,354 @@
+# %% [markdown]
+# # Using deep Gaussian processes with GPflux for Bayesian optimization.
+
+# %%
+import numpy as np
+import tensorflow as tf
+
+np.random.seed(1794)
+tf.random.set_seed(1794)
+tf.keras.backend.set_floatx("float64")
+
+# %% [markdown]
+# ## Describe the problem
+# In this example, we look to find the minimum value of the two-dimensional Branin function over the hypercube $[0, 1]^2$, modified with a discontinuity. We can represent the search space using a `Box`, and plot contours of the Branin over this space.
+#
+#
+
+# %%
+import gpflow
+from trieste.objectives import (
+    michalewicz,
+    MICHALEWICZ_2_MINIMUM,
+    MICHALEWICZ_5_MINIMUM,
+)
+from trieste.objectives.utils import mk_observer
+from util.plotting_plotly import plot_function_plotly
+from trieste.space import Box
+from math import pi
+
+function = michalewicz
+F_MINIMIZER = MICHALEWICZ_2_MINIMUM
+
+search_space = Box([0, 0], [pi, pi])
+
+fig = plot_function_plotly(
+    function,
+    search_space.lower,
+    search_space.upper,
+    grid_density=100
+)
+fig.update_layout(height=400, width=400)
+fig.show()
+
+# %% [markdown]
+# ## Sample the observer over the search space
+#
+# Sometimes we don't have direct access to the objective function. We only have an observer that indirectly observes it. In _Trieste_, an observer can output a number of datasets. In our case, we only have one dataset, the objective. We can convert a function with `scaled_discts_branin`'s signature to a single-output observer using `mk_observer`.
+#
+# The optimization procedure will benefit from having some starting data from the objective function to base its search on. We sample a five point space-filling design from the search space and evaluate it with the observer. For continuous search spaces, Trieste supports random, Sobol and Halton initial designs.
+
+# %%
+import trieste
+
+observer = mk_observer(function)
+
+num_initial_points = 10
+num_acquisitions = 20
+initial_query_points = search_space.sample_sobol(num_initial_points)
+initial_data = observer(initial_query_points)
+
+# %% [markdown]
+# ## Model the objective function
+#
+# The Bayesian optimization procedure estimates the next best points to query by using a probabilistic model of the objective. We'll use a two layer deep Gaussian process (DGP), built using GPflux. We also compare to a (shallow) GP.
+
+# %%
+from trieste.models.gpflux import GPfluxModelConfig, build_vanilla_deep_gp
+from gpflow.utilities import set_trainable
+
+
+def build_dgp_model(data):
+    variance = tf.math.reduce_variance(data.observations)
+
+    dgp = build_vanilla_deep_gp(data.query_points, num_layers=2, num_inducing=100)
+    dgp.f_layers[-1].kernel.kernel.variance.assign(variance)
+    dgp.f_layers[-1].mean_function = gpflow.mean_functions.Constant()
+    dgp.likelihood_layer.likelihood.variance.assign(1e-5)
+    set_trainable(dgp.likelihood_layer.likelihood.variance, False)
+
+    epochs = 200
```

Comment:
also worth pointing out we use SGD and Adam to train it, perhaps also mentioning Kereas fit method